v2/yarpctchannel/inbound.go

```diff
@@ -18,79 +18,123 @@
 }
 
-// NewInbound returns a new TChannel inbound backed by a shared TChannel
-// transport.
-// There should only be one inbound for TChannel since all outbounds send the
-// listening port over non-ephemeral connections so a service can deduplicate
-// locally- and remotely-initiated persistent connections.
-func (t *Transport) NewInbound() *Inbound {
-	return &Inbound{
-		once:      lifecycle.NewOnce(),
-		transport: t,
+// Start starts this Inbound.
+func (i *Inbound) Start(_ context.Context) error {
+
+	if i.Logger == nil {
+		i.Logger = zap.NewNop()
 	}
-}
 
-// SetRouter configures a router to handle incoming requests.
-// This satisfies the transport.Inbound interface, and would be called
-// by a dispatcher when it starts.
-func (i *Inbound) SetRouter(router transport.Router) {
-	i.transport.router = router
-}
+	if i.Tracer == nil {
+		i.Tracer = opentracing.GlobalTracer()
+	}
 
-// Transports returns a slice containing the Inbound's underlying
-// Transport.
-func (i *Inbound) Transports() []transport.Transport {
-	return []transport.Transport{i.transport}
-}
+	// Create a TChannel
+	chopts := tchannel.ChannelOptions{
+		Tracer: i.Tracer,
+		Handler: handler{
+			router:     i.Router,
+			tracer:     i.Tracer,
+			logger:     i.Logger,
+			headerCase: i.HeaderCase,
+		},
+	}
+	ch, err := tchannel.NewChannel(i.Service, &chopts)
+	if err != nil {
+		return err
+	}
+	i.ch = ch
 
-// Start starts this Inbound. Note that this does not start listening for
-// connections; that occurs when you start the underlying ChannelTransport is
-// started.
-func (i *Inbound) Start() error {
-	return i.once.Start(func() error {
-		i.transport.logger.Info("started TChannel inbound", zap.String("address", i.transport.addr))
-		if i.transport.router == nil || len(i.transport.router.Procedures()) == 0 {
-			i.transport.logger.Warn("no procedures specified for tchannel inbound")
+	// Create a listen address
+	if i.Addr == "" {
+		addr, err := tchannel.ListenIP()
+		if err != nil {
+			return err
 		}
-		return nil
-	})
-}
-
-// Stop stops the TChannel outbound. This currently does nothing.
-func (i *Inbound) Stop() error {
-	return i.once.Stop(nil)
-}
+		i.Addr = addr.String()
+	}
 
-// IsRunning returns whether the Inbound is running.
-func (i *Inbound) IsRunning() bool {
-	return i.once.IsRunning()
-}
+	if i.Listener == nil {
+		listener, err := net.Listen("tcp", i.Addr)
+		if err != nil {
+			return err
+		}
+		i.Listener = listener
+	}
 
-// Introspect returns the state of the inbound for introspection purposes.
-func (i *Inbound) Introspect() introspection.InboundStatus {
-	stateString := ""
-	if i.transport.ch != nil {
-		stateString = i.transport.ch.State().String()
+	err = i.ch.Serve(i.Listener)
+	if err != nil {
+		return err
 	}
-	return introspection.InboundStatus{
-		Transport: "tchannel",
-		Endpoint:  i.transport.addr,
-		State:     stateString,
+
+	i.Logger.Info("started TChannel inbound", zap.String("address", i.Listener.Addr().String()))
```

Comment:
Don't need to call `.String()` here. Can use [`zap.Stringer`](https://godoc.org/go.uber.org/zap#Stringer) instead.

    zap.Stringer("address", i.listener.Addr())